[GTK] Add WebKitClipboardPermissionRequest to handle DOM paste access requests

carlosgcampos opened the pull request:
#### 2ec3bfdc8377ed6d94254d62ccd5f53f33e47d58
<pre>
[GTK] Add WebKitClipboardPermissionRequest to handle DOM paste access requests
<a href="https://bugs.webkit.org/show_bug.cgi?id=254510">https://bugs.webkit.org/show_bug.cgi?id=254510</a>

Reviewed by Michael Catanzaro and Adrian Perez de Castro.

* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/UIProcess/API/glib/WebKitClipboardPermissionRequest.cpp: Added.
(webkitClipboardPermissionRequestAllow):
(webkitClipboardPermissionRequestDeny):
(webkit_permission_request_interface_init):
(webkitClipboardPermissionRequestDispose):
(webkit_clipboard_permission_request_class_init):
(webkitClipboardPermissionRequestCreate):
* Source/WebKit/UIProcess/API/glib/WebKitClipboardPermissionRequest.h.in: Added.
* Source/WebKit/UIProcess/API/glib/WebKitClipboardPermissionRequestPrivate.h: Added.
* Source/WebKit/UIProcess/API/glib/webkit.h.in:
* Source/WebKit/UIProcess/API/gtk/PageClientImpl.cpp:
(WebKit::PageClientImpl::requestDOMPasteAccess):
* Tools/MiniBrowser/gtk/BrowserTab.c:
(permissionRequestDialogResponse):
(decidePermissionRequest):
(browser_tab_class_init):
* Tools/TestWebKitAPI/Tests/WebKitGtk/TestWebViewEditor.cpp:
(Clipboard::writeText):
(testWebViewClipboardPermissionRequest):
(beforeAll):

Canonical link: <a href="https://commits.webkit.org/262202@main">https://commits.webkit.org/262202@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5e66b13784c020d9c7003d98b37d2197fac8e1a2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/907 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/930 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/966 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/1307 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/802 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/894 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/980 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1015 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1009 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/914 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/854 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/849 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/1232 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/905 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/846 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/843 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/821 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/873 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/1872 "267 api tests failed or timed out") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/862 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/816 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/809 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/857 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/876 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/93 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitClipboardPermissionRequest.cpp:
```diff
@@ -0,0 +1,101 @@
+ * WebKitClipboardPermissionRequest:
+ * @See_also: #WebKitPermissionRequest, #WebKitWebView
+ *
+ * A permission request for reading clipboard contents.
```
We'll never want to have a permission request for writing to the clipboard, then? That will just always be allowed no matter what?

carlosgcampos replied:
I don't know, but for now reading is the only operation requiring permission. We can change it later if needed.

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/PageClientImpl.cpp:
```diff
@@ -552,13 +554,19 @@ void PageClientImpl::derefView()
         }
 
-        completionHandler(DOMPasteAccessResponse::DeniedForGesture);
+        if (!WEBKIT_IS_WEB_VIEW(weakWebView.get())) {
```
Shouldn't this be:

```suggestion
        if (!weakWebView.get()) {
```

? It's either a WebKitWebView or it's nullptr, right?

carlosgcampos replied:
No, it could be WebKitWebViewBase which doesn't support permission requests.

mcatanzaro reviewed (approved):
Good idea. API looks fine. Nice test.

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitClipboardPermissionRequest.cpp:
```diff
@@ -0,0 +1,101 @@
+    WebKitClipboardPermissionRequestPrivate* priv = WEBKIT_CLIPBOARD_PERMISSION_REQUEST(request)->priv;
+
+    if (auto completionHandler = std::exchange(priv->completionHandler, nullptr))
+        completionHandler(WebCore::DOMPasteAccessResponse::GrantedForGesture);
```
You don't need the `std::exchange()`, the [overloaded call operator](https://github.com/WebKit/WebKit/blob/ee0988b18d7348a73ce7db08beafbaf92788a627/Source/WTF/wtf/CompletionHandler.h#L75) already does it.

```suggestion
    if (priv->completionHandler)
        priv->completionHandler(WebCore::DOMPasteAccessResponse::GrantedForGesture);
```

carlosgcampos replied:
Ok.

mcatanzaro replied:
Unfortunately the `std::exchange` was needed here. Now it will assert with "Completion handler should not be called more than once." This requires follow-up.

aperezdc replied:
But that means that it's getting called twice, signalling a bug elsewhere. Completion handlers should not be called more than once after all!

mcatanzaro replied:
The bug is right here. It's going to be called multiple times because you told Carlos to remove this `std::exchange` that was needed to prevent that. It's called for the first time when the application denies the permission request, then called again at least once, but potentially many times, in dispose, once for each call to dispose.

aperezdc replied:
After the first call to `CompletionHandler::operator()`, its `m_function` member will be set to `nullptr` (that's the `std::exchange` inside the operator) and from that point onwards the `operator bool()` overload ([just two lines above](https://github.com/WebKit/WebKit/blob/ee0988b18d7348a73ce7db08beafbaf92788a627/Source/WTF/wtf/CompletionHandler.h#L69)) will return `false`, which means the `if (priv->completionHandler)` check is `false` and this code will _not_ be using the operator again. The bug has to be elsewhere.

mcatanzaro replied:
Ohhhh... OK then, you are correct. Huh.

I didn't actually notice this crashing. Was just looking at the code. Today I learned. ;)

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitClipboardPermissionRequest.cpp:
```diff
@@ -0,0 +1,101 @@
+    WebKitClipboardPermissionRequestPrivate* priv = WEBKIT_CLIPBOARD_PERMISSION_REQUEST(request)->priv;
+
+    if (auto completionHandler = std::exchange(priv->completionHandler, nullptr))
+        completionHandler(WebCore::DOMPasteAccessResponse::DeniedForGesture);
```
Same here:

```suggestion
    if (priv->completionHandler, nullptr)
        priv->completionHandler(WebCore::DOMPasteAccessResponse::DeniedForGesture);
```

aperezdc reviewed (approved):
Other than a couple of small nits, patch LGTM if you address @mcatanzaro's comments.

carlosgcampos committed:
[GTK] Add WebKitClipboardPermissionRequest to handle DOM paste access requests
https://bugs.webkit.org/show_bug.cgi?id=254510

Reviewed by Michael Catanzaro and Adrian Perez de Castro.

* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/UIProcess/API/glib/WebKitClipboardPermissionRequest.cpp: Added.
(webkitClipboardPermissionRequestAllow):
(webkitClipboardPermissionRequestDeny):
(webkit_permission_request_interface_init):
(webkitClipboardPermissionRequestDispose):
(webkit_clipboard_permission_request_class_init):
(webkitClipboardPermissionRequestCreate):
* Source/WebKit/UIProcess/API/glib/WebKitClipboardPermissionRequest.h.in: Added.
* Source/WebKit/UIProcess/API/glib/WebKitClipboardPermissionRequestPrivate.h: Added.
* Source/WebKit/UIProcess/API/glib/webkit.h.in:
* Source/WebKit/UIProcess/API/gtk/PageClientImpl.cpp:
(WebKit::PageClientImpl::requestDOMPasteAccess):
* Tools/MiniBrowser/gtk/BrowserTab.c:
(permissionRequestDialogResponse):
(decidePermissionRequest):
(browser_tab_class_init):
* Tools/TestWebKitAPI/Tests/WebKitGtk/TestWebViewEditor.cpp:
(Clipboard::writeText):
(testWebViewClipboardPermissionRequest):
(beforeAll):

Canonical link: https://commits.webkit.org/262202@main
